### PR TITLE
Post about Prometheus Health Checker

### PR DESCRIPTION
**Pull request by tsx**

### CODE REVIEW

File: _posts/2021-09-30-prometheus-health-check-bin/index.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+We also would not want to rely on the process to self-report any issues. If it’s blocked by something, there’s a good chance that an attempt to self-report from within would also be blocked.
+
+## Reporting regular progress
+
+The main indication of a process being stuck is actually *absence* of normal work instead of some active action. We changed the process to report a timestamp of when it processed an item last time. This way, we can detect a stuck process much earlier - by last reported timestamp staying the same as real time goes forward.
+
+```
+# HELP closeio_last_iteration_ts Number of subscriptions monitored by DMS
+# TYPE closeio_last_iteration_ts gauge
+closeio_last_iteration_ts 1633021679.8027766
```
Comment (philfreo) [inline _posts/2021-09-30-prometheus-health-check-bin/index.md:29]: Is this Prometheus syntax? You haven't mentioned that anywhere yet.

File: _posts/2021-09-30-prometheus-health-check-bin/index.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
+---
+layout: post
+title: 'The Right Tool for the Job'
```
Comment (philfreo) [inline _posts/2021-09-30-prometheus-health-check-bin/index.md:3]: I suggest using a more descriptive title

File: _posts/2021-09-30-prometheus-health-check-bin/index.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+date: 2021-09-30
+permalink: /posts/prometheus-health-check-bin/
+author: Vyacheslav Tverskoy
+thumbnail: ''
+metaDescription: ''
+tags: [Engineering, Open Source]
```
Comment (philfreo) [inline _posts/2021-09-30-prometheus-health-check-bin/index.md:9]: perhaps Monitoring or some other slightly more specific tag?

**Review by philfreo — APPROVED** [review]
great stuff!

File: _posts/2021-09-30-prometheus-health-check-bin/index.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+We also would not want to rely on the process to self-report any issues. If it’s blocked by something, there’s a good chance that an attempt to self-report from within would also be blocked.
+
+## Reporting regular progress
+
+The main indication of a process being stuck is actually *absence* of normal work instead of some active action. We changed the process to report a timestamp of when it processed an item last time. This way, we can detect a stuck process much earlier - by last reported timestamp staying the same as real time goes forward.
```
Comment (jkemp101) [inline _posts/2021-09-30-prometheus-health-check-bin/index.md:24]: Change
>it processed an item last time

to

>it last processed an item

or just remove `time`?

File: _posts/2021-09-30-prometheus-health-check-bin/index.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+Now it’s time to make the system self-heal.
+
+## Health checks
+
+Most of the compute workloads we run at Close are orchestrated with Kubernetes. It’s useful not only for deploying our app multiple times a day as we ship new features and improve internals, but also adds a lot of resiliency when operating in a cloud-based environment. It can restart a process or a container when it exits abnormally or appears non-functional.
+
+In our case, the process did looked somewhat functional since it responded to a basic health check and was able to report current metric values, despite not doing any useful work. The next task is to actually utilize the metrics to recognize that a specific stale timestamp metric means that process is not healthy anymore and we need it restarted.
```
Comment (jkemp101) [inline _posts/2021-09-30-prometheus-health-check-bin/index.md:40]: >did looked somewhat

to

>did look somewhat

File: _posts/2021-09-30-prometheus-health-check-bin/index.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+As part of the existing health check, we already had a simple shell script that used an existing metric. Here it is in its entirety:
+
+```bash
+#!/bin/sh
+curl -s http://127.0.0.1:$1/metrics/ | grep 'cio_service_state 1.0'
+```
+
+All it does is query the metrics endpoint via http and grepping for a predefined string. It works fine for the most part, but it’s not as easy to extend this to check *freshness* of some different metric. To do that in a shell script, you’d hold onto the metrics output, grep it multiple times and do arithmetic on timestamps. With shell scripts, any kind of logic or control flow beyond simple linear command chains becomes hard to maintain, and correctly dealing with all the ways things could go wrong is even harder.
```
Comment (jkemp101) [inline _posts/2021-09-30-prometheus-health-check-bin/index.md:49]: `grepping` to `grep`

**Review by jkemp101 — APPROVED** [review]
Just a few things I missed from my first review

File: _posts/2021-09-30-prometheus-health-check-bin/index.md (hunk 1 above)
Comment (lmickh) [reply _posts/2021-09-30-prometheus-health-check-bin/index.md:29]: Yes, this is what hitting a prometheus metrics endpoint looks like.

Comment (tsx) [reply _posts/2021-09-30-prometheus-health-check-bin/index.md:29]: Adding a sentence to clarify what it is.

**Pull request merged**